### PR TITLE
ANA-1190: Adding RFR/CAD Swap Examples

### PR DESCRIPTION
**Pull request by njmfinbourne**
# Pull Request Checklist

- [x] Read the [contributing guidelines](../blob/master/docs/CONTRIBUTING.md)
- [ ] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

Describe the code changes for the reviewers, explain the solution you have provided and how it fixes the issue

### CODE REVIEW

File: sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs [hunk 1]
```diff
@@ -36,8 +40,16 @@ public static class TestDataUtilities
         public const string LusidCashIdentifier = "Instrument/default/Currency";
         public const string LusidInstrumentIdentifier = "Instrument/default/LusidInstrumentId";
 
-        public static DateTimeOffset EffectiveAt = new DateTimeOffset(2020, 1, 2, 0, 0, 0, TimeSpan.Zero);
-
+        public static DateTimeOffset StartDate = new DateTimeOffset(2020, 1, 2, 0, 0, 0, TimeSpan.Zero);
+        public static DateTimeOffset EffectiveAt = StartDate;
+        public static DateTimeOffset ResetDate = StartDate.AddYears(-1); // Temp reset date -> will change examples to pull resets from instruments when implemented
+        
+        public static List<decimal> exampleRates1 = new List<decimal> { 1.0m, 0.992548449440757m, 0.985152424487251m, 0.977731146620901m, 0.970365774179742m, 0.85m };
```
Comment (jz-finbourne) [inline sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs:47]: (variable naming) What rates are these, discount curve rates?

Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs:47]: Agreed not clear - will update.

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/EquitySwapExamples.cs [hunk 2]
```diff
@@ -22,37 +22,44 @@ protected override void CreateAndUpsertMarketDataToLusid(string scope, ModelSele
...
                 135m,
                 "USD",
-                TestDataUtilities.EffectiveAt);
-            quotesToUpsert.Add(equityUnderlyingQuoteRequestAtEffectiveAt.QuoteId.ToString(), equityUnderlyingQuoteRequestAtEffectiveAt);
+                TestDataUtilities.EffectiveAt,
+                QuoteSeriesId.QuoteTypeEnum.Price);
             
             // FOR lifecycle example, we value around maturity of the equity swap on the following days
-            var days = new []{19, 20, 21, 22, 23};
+            var days = new []{0, 1, 2, 3, 4};
```
Comment (SamuelColvinFinbourne) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/EquitySwapExamples.cs:37]: Why did you need to change the days being tested on?

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapWithNamedConventionsExamples.cs [hunk 3]
```diff
@@ -20,15 +20,32 @@ protected override void CreateAndUpsertMarketDataToLusid(string scope, ModelSele
             // For LUSID to pick up these quotes, we have added a RIC rule to the recipe (see BuildRecipeRequest in TestDataUtilities.cs) 
             // The RIC rule has a large quote interval, this means that we can use one reset quote for all the resets.
             // For accurate pricing, one would want to upsert a quote per reset. 
-            var quoteRequest = TestDataUtilities.BuildQuoteRequest("USD6M", QuoteSeriesId.InstrumentIdTypeEnum.ClientInternal, 0.05m, "USD", TestDataUtilities.EffectiveAt);
+            
+            var d =  new DateTimeOffset(2019, 12, 31, 0, 0, 0, TimeSpan.Zero);
```
Comment (SamuelColvinFinbourne) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapWithNamedConventionsExamples.cs:24]: "effectiveDate"? why this special one, and not a generic one?

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            // The RIC rule has a large quote interval, this means that we can use one reset quote for all the resets.
+            // For accurate pricing, one would want to upsert a quote per reset.
+            InterestRateSwap irs = instrument as InterestRateSwap;
+
+            FloatingLeg floatLeg =
+                (FloatingLeg) irs.Legs.Where(x => x.InstrumentType == LusidInstrument.InstrumentTypeEnum.FloatingLeg).First();
+
+            var indexName = floatLeg.LegDefinition.IndexConvention.IndexName;
+
+            UpsertQuotesResponse upsertResponse;
```
Comment (SamuelColvinFinbourne) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:30]: Why instantiate here?

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            string recipeCode,
+            string instrumentID)
+        {
+            var swap = (InterestRateSwap) instrument;
+            var cashflows = _transactionPortfoliosApi.GetPortfolioCashFlows(
+                scope: scope,
+                code: portfolioCode,
+                effectiveAt: TestDataUtilities.EffectiveAt.AddYears(3),
+                windowStart: swap.StartDate.AddDays(-3),
+                windowEnd: swap.MaturityDate.AddDays(3),
```
Comment (SamuelColvinFinbourne) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:113]: Maybe add a few more days, or just make it a month. You don't want to accidently miss any cashflows because of roll conventions. It would be a pretty hard bug to spot

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            var getResponse = _instrumentsApi.GetInstruments("ClientInternal", new List<string> { uniqueId });
+            ValidateInstrumentResponse(getResponse, uniqueId);
+            
+            var retrieved = getResponse.Values.First().Value.InstrumentDefinition;
+            Assert.That(retrieved.InstrumentType == LusidInstrument.InstrumentTypeEnum.InterestRateSwap);
+            var roundTripSwap = retrieved as InterestRateSwap;
+            Assert.That(roundTripSwap, Is.Not.Null);
+            Assert.That(roundTripSwap.MaturityDate, Is.EqualTo(swap.MaturityDate));
+            Assert.That(roundTripSwap.StartDate, Is.EqualTo(swap.StartDate));
+            Assert.That(roundTripSwap.Legs.Count, Is.EqualTo(swap.Legs.Count));
```
Comment (SamuelColvinFinbourne) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:154]: Why not check some data that distinguishes the type of swap being upserted, like IndexName?

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        [TestCase(ModelSelection.ModelEnum.Discounting, "SOFR")]
+        [TestCase(ModelSelection.ModelEnum.ConstantTimeValueOfMoney, "ESTR")]
+        [TestCase(ModelSelection.ModelEnum.Discounting, "ESTR")]
+        [TestCase(ModelSelection.ModelEnum.ConstantTimeValueOfMoney, "SONIA")]
+        [TestCase(ModelSelection.ModelEnum.Discounting, "SONIA")]
+        [TestCase(ModelSelection.ModelEnum.ConstantTimeValueOfMoney, "TONA")]
+        [TestCase(ModelSelection.ModelEnum.Discounting, "TONA")]
+        [TestCase(ModelSelection.ModelEnum.ConstantTimeValueOfMoney, "SARON")]
+        [TestCase(ModelSelection.ModelEnum.Discounting, "SARON")]
+        public void InterestRateSwapRfRPortfolioCashFlowsExample(ModelSelection.ModelEnum model, string rfr)
```
Comment (SamuelColvinFinbourne) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:188]: "rfr" doesn't mean anything to me. Maybe a more verbose name.

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapWithNamedConventionsExamples.cs (hunk 3 above)
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapWithNamedConventionsExamples.cs:24]: Will use the generic resetDate.

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs (hunk 7 above)
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:188]: RFR is quite a common term like LIBOR - don't think it needs expanding in this context.

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs (hunk 6 above)
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:154]: Same comment for other examples - comes back to question of the purpose of this SDK code. Examples or actual tests.... My view is that they are examples rather than complete tests (tests should exist in Lusid rather than be caught here).

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs (hunk 5 above)
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:113]: This is an example of how to get cashflows - the window is arbitrary and might only be used to extract specific cashflows - doesn't necessarily need to get all cashflows.

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs (hunk 4 above)
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:30]: Was a hangover from a previous iteration of the code - will remove.

File: sdk/Lusid.Sdk.Tests/Utilities/InstrumentExamples.cs [hunk 8]
```diff
@@ -138,40 +139,18 @@ internal static LusidInstrument CreateExampleZeroCouponBond()
...
-                    fixedLeg
-                },
-                instrumentType: LusidInstrument.InstrumentTypeEnum.InterestRateSwap
-            );
-        }
-
-        private static IndexConvention CreateExampleIndexConventions(string code)
+        private static IndexConvention CreateExampleIndexConventions(string currency, string indexName, string tenor, string dayCount, string fixingRef)
         {
+            string code = currency + indexName + tenor;
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/InstrumentExamples.cs:145]: This was how it was created previously - there is no way at this stage to know what valuation type is being called. Will put a comment to this effect.

File: sdk/Lusid.Sdk.Tests/Utilities/InstrumentExamples.cs [hunk 9]
```diff
@@ -67,46 +68,46 @@ internal static LusidInstrument CreateExampleEquityOption(bool isCashSettled = f
...
-                instrumentType: LusidInstrument.InstrumentTypeEnum.Equity, 
-                domCcy: "USD", 
-                identifiers: new EquityAllOfIdentifiers(isin: "US-000402625-0") 
+                instrumentType: LusidInstrument.InstrumentTypeEnum.Equity,
+                domCcy: "USD",
+                identifiers: new EquityAllOfIdentifiers(isin: "US-000402625-0")
             );
 
-        private static FlowConventions CreateExampleFlowConventions()
+        private static FlowConventions CreateExampleFlowConventions(string currency, string paymentFrequency, string rollConvention, string dayCount, int settleDays, int resetDays)
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/InstrumentExamples.cs:84]: For consistency I will leave as is.

File: sdk/Lusid.Sdk.Tests/Utilities/DemoInstrumentBase.cs [hunk 10]
```diff
@@ -119,12 +119,16 @@ internal void CallLusidGetPortfolioCashFlowsEndpoint(LusidInstrument instrument,
                 // SimpleStatic pricing is lookup pricing. As such, we upsert a quote.
                 // Note that inside CreatePortfolioAndInstrument, the method TestDataUtilities.BuildInstrumentUpsertRequest books the instrument using "ClientInternal".
                 // Hence upsert a quote using ClientInternal as the instrumentIdType.
-                var quoteRequest = TestDataUtilities.BuildQuoteRequest(
+                var quoteRequest = new Dictionary<string, UpsertQuoteRequest>();
+                TestDataUtilities.BuildQuoteRequest(
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/DemoInstrumentBase.cs:123]: Yes - not huge use here but in other cases where mutliple quoterequests are made then it is useful.

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs (hunk 6 above)
Comment (SamuelColvinFinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:153]: Fair enough, I would argue then that this is already too many assertions, but it will be fine as is.

File: sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs [hunk 11]
```diff
@@ -213,61 +222,45 @@ public static PortfolioHolding BuildCashPortfolioHolding(string currency, string
...
-                    upsertQuoteRequests.Add($"day_{days}_inverseFx_rate", inverseFxRate);
-                }
+                var rateAddOn = (useConstantFxRate) ? rate : rate + days;
+                BuildQuoteRequest(
+                    upsertQuoteRequests,
+                    $"day_{days}_fx_rate",
+                    $"{ccy1}/{ccy2}",
+                    QuoteSeriesId.InstrumentIdTypeEnum.CurrencyPair,
+                    rateAddOn,
+                    $"{ccy2}",
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs:244]: Are you sure - when you upsert the quote that field is used in the column labelled Currency (when looking at quotes)

File: sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs [hunk 12]
```diff
@@ -282,132 +275,69 @@ public static PortfolioHolding BuildCashPortfolioHolding(string currency, string
...
 
             return upsertQuoteRequests;
         }
 
-
         /// <summary>
         /// Helper method to create simple equity or FX quote upsert request
         /// </summary>
-        internal static UpsertQuoteRequest BuildSimpleQuoteUpsertRequest(
+        internal static UpsertQuoteRequest BuildSimpleQuoteUpsertRequestOld(
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs:294]: should be removed

File: sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs [hunk 13]
```diff
@@ -592,28 +536,32 @@ private static string GetMarketAsset(LusidInstrument option, MarketQuote.QuoteTy
...
                 MarketDataKeyRule.QuoteTypeEnum.Price,
                 field: "mid",
                 quoteInterval: "10Y");
+                                     
 
             // projection rule here is used by Lusid to locate the rate curves for projections of index rates
             // (e.g. projected LIBOR rates)
             // Used when pricing instruments that depend on an IndexConvention.
             // For instance, interest rate swaps and options.
+            // Curves should be upserted with the MarketAsset of the form {ccy}.{tenor}.{indexname} e.g. USD.1D.SOFR or USD.6M.LIBOR
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs:556]: Same as discounting - these are curves.  There is a comment at the top which specifies that this is for projection curves.

File: sdk/Lusid.Sdk.Tests/tutorials/Ibor/StructuredResultStore.cs [hunk 14]
```diff
@@ -628,20 +628,28 @@ public void TestFindOrCalculate_PortfolioResult()
             _transactionPortfoliosApi.UpsertTransactions(scope, portfolioCode, transactionRequest2);
             
             // Create and upsert quotes, one for each instrument defined above
-            var quoteRequest1 = TestDataUtilities.BuildQuoteRequest(
+            var quoteRequest1 = new Dictionary<string, UpsertQuoteRequest>();
+            TestDataUtilities.BuildQuoteRequest(
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Ibor/StructuredResultStore.cs:632]: Yes - should just be one that is added to.

File: sdk/Lusid.Sdk.Tests/Utilities/InstrumentExamples.cs [hunk 15]
```diff
@@ -1,12 +1,13 @@
 using System;
 using System.Collections.Generic;
+using System.Security;
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/InstrumentExamples.cs:3]: Have removed

File: sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs [hunk 16]
```diff
@@ -38,7 +38,16 @@ public static class TestDataUtilities
         public const string LusidCashIdentifier = "Instrument/default/Currency";
         public const string LusidInstrumentIdentifier = "Instrument/default/LusidInstrumentId";
 
-        public static DateTimeOffset EffectiveAt = new DateTimeOffset(2020, 1, 2, 0, 0, 0, TimeSpan.Zero);
+        public static DateTimeOffset StartDate = new DateTimeOffset(2020, 1, 2, 0, 0, 0, TimeSpan.Zero);
+        public static DateTimeOffset EffectiveAt = StartDate;
+        public static DateTimeOffset ResetDate = StartDate.AddYears(-1); // Temp reset date -> will change examples to pull resets from instruments when implemented
+
+        public static List<decimal> exampleDiscountFactors1 = new List<decimal> { 1.0m, 0.992548449440757m, 0.985152424487251m, 0.977731146620901m, 0.970365774179742m, 0.85m };
+        public static List<decimal> exampleDiscountFactors2 = new List<decimal> { 1.0m, 0.995026109593975m, 0.990076958773721m, 0.985098445011387m, 0.980144965261876m, 0.9m };
```
[the remark this answers is not in the document]
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/Utilities/TestDataUtilities.cs:46]: updated

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+            // CAN NOW UPSERT TO LUSID
+            var uniqueId = swap.InstrumentType + Guid.NewGuid().ToString(); 
+            var instrumentsIds = new List<(LusidInstrument, string)>{(swap, uniqueId)};
+            var definitions = TestDataUtilities.BuildInstrumentUpsertRequest(instrumentsIds);
+            
+            var upsertResponse = _instrumentsApi.UpsertInstruments(definitions);
+            ValidateUpsertInstrumentResponse(upsertResponse);
+
+            // CAN NOW QUERY FROM LUSID
+            var getResponse = _instrumentsApi.GetInstruments("ClientInternal", new List<string> { uniqueId });
```
Comment (rizwansaeed) [inline sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:144]: you may need to qualify the `GetInstruments` with the `asAt` returned from `UpsertInstruments` do avoid any DEI eventual consistency issues

**Review by rizwansaeed — CHANGES_REQUESTED** [review]

File: sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs (hunk 17 above)
Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:144]: There shouldn't be an issue as the uniqueID is genrated during this the test and only one version is upserted. The asat is optional:
<param name="asAt">The asAt datetime at which to retrieve the instrument definitions. Defaults to returning the latest version of each instrument definition if not specified. (optional)

Can't see there being any conflicts when this test is run - even if run in parallel?

Comment (rizwansaeed) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:144]: its less about the uniqueness of the id and more to do with whether the instrument will be returned from the list call. It _may_ not get returned if it hasn't been indexed, but specifying the `asAt` on the list will guarantee it is returned

Comment (njmfinbourne) [reply sdk/Lusid.Sdk.Tests/tutorials/Instruments/InterestRateSwapRfRExamples.cs:144]: Ok - will need to change all the tests that use this.

**Review by rizwansaeed — APPROVED** [review]
